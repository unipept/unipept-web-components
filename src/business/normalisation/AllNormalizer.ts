import Normalizer from "./Normalizer";

/**
 * Normalises values in a grid relative to one absolute maximum and minimum value for the whole grid. The largest and
 * smallest value for the complete grid are selected after which all other values are normalised according to these
 * values.
 *
 * @author Pieter Verschaffelt
 */
export default class AllNormalizer implements Normalizer {
    public normalize(data: number[][]): number[][] {
        if (data.length === 0 || data[0].length === 0) {
            return data;
        }

        let min = Infinity;
        let max = -Infinity;

        for (const row of data) {
            for (const value of row) {
                min = Math.min(min, value);
                max = Math.max(max, value);
            }
        }

        const output: number[][] = [];

        for (const row of data) {
            const newRow: number[] = [];
            for (const value of row) {
                if (max - min !== 0) {
                    newRow.push((value - min) / (max - min));
                } else {
                    newRow.push(0);
                }
            }
            output.push(newRow);
        }

        return output;
    }

    public toString(): string {
        return "AllNormalizer";
    }
}
