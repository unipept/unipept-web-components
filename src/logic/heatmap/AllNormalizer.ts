import { Normalizer } from "./Normalizer";

export default class AllNormalizer implements Normalizer {
    public normalize(data: number[][]): number[][] {
        if (data.length === 0 || data[0].length === 0) {
            return data;
        }

        let min = Infinity;
        let max = -Infinity;

        for (const row of data) {
            for (const value of row) {
                min = Math.min(min, value);
                max = Math.max(max, value);
            }
        }

        const output: number[][] = [];

        for (const row of data) {
            const newRow: number[] = [];
            for (const value of row) {
                if (max - min !== 0) {
                    newRow.push((value - min) / (max - min));
                } else {
                    newRow.push(0);
                }
            }
            output.push(newRow);
        }

        return output;
    }
}
