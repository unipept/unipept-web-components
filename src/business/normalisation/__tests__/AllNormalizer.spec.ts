import AllNormalizer from "@/business/normalisation/AllNormalizer";

describe("AllNormalizer", () => {
    it("normalizes a complete grid correctly", () => {
        const normalizer = new AllNormalizer();
        const data = [
            [4, 4, 2, 1],
            [1, 2, 1, 0],
            [4, 3, 0, 4]
        ];

        const normalizedResult = normalizer.normalize(data);
        expect(normalizedResult).toMatchSnapshot();
    });

    it("normalizes a grid with negative values correctly", () => {
        const normalizer = new AllNormalizer();
        const data = [
            [4, 4, -2, 1],
            [-1, 2, 1, 0],
            [-4, 3, 0, 4]
        ];

        const normalizedResult = normalizer.normalize(data);
        // There should no longer be negative values in the result, since everything is mapped onto the [0, 1] interval.
        expect(normalizedResult).toMatchSnapshot();
    });

    it("does not crash on all-zero grids", () => {
        const normalizer = new AllNormalizer();
        const data = [
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0]
        ];

        const normalizedResult = normalizer.normalize(data);
        // There should no longer be negative values in the result, since everything is mapped onto the [0, 1] interval.
        expect(normalizedResult).toMatchSnapshot();
    });
});
