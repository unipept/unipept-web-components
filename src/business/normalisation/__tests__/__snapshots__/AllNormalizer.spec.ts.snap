// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllNormalizer does not crash on all-zero grids 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`AllNormalizer normalizes a complete grid correctly 1`] = `
Array [
  Array [
    1,
    1,
    0.5,
    0.25,
  ],
  Array [
    0.25,
    0.5,
    0.25,
    0,
  ],
  Array [
    1,
    0.75,
    0,
    1,
  ],
]
`;

exports[`AllNormalizer normalizes a grid with negative values correctly 1`] = `
Array [
  Array [
    1,
    1,
    0.25,
    0.625,
  ],
  Array [
    0.375,
    0.75,
    0.625,
    0.5,
  ],
  Array [
    0,
    0.875,
    0.5,
    1,
  ],
]
`;
